### PR TITLE
Move type serialization to emitter

### LINKED ISSUE #3663: Decorators and SystemJS
```ts
// @experimentaldecorators: true
// @emitdecoratormetadata: true
// @target: system

import { Foo } from 'foo';
declare var decorator: any;

@decorator
export class Test {
	constructor(private foo: Foo) { }
}
```

```js
// ...
System.register(['foo'], function(exports_1) {
    var Foo_1;
    var Test;
    return {
        setters:[
            function (_Foo_1) {
                Foo_1 = _Foo_1;
            }],
        execute: function() {
            Test = (function () {
                function Test(foo) {
                    this.foo = foo;
                }
                Test = __decorate([
                    decorator, 
                    __metadata('design:paramtypes', [Foo])
                ], Test);
                return Test;
            })();
            exports_1("Test", Test);
        }
    }
});
```

The problem is with `__metadata('design:paramtypes', [Foo])`, at runtime `Foo` doesn't exist.  I'm guessing `Foo_1` was intended here?

### PR DESCRIPTION
**Pull request by rbuckton**
Moves the logic for decorator metadata type serialization to the emitter.

This follows #3380, and also fixes #3663.

### CODE REVIEW

**Comment by gdi2290** [comment]
:+1:

**Comment by DanielRosenwasser** [comment]
@gdi2290 1.5 is basically stabilizing and not taking anything that's not extremely high priority. This will be going into 1.6.

**Comment by gdi2290** [comment]
we found a workaround @DanielRosenwasser so no worries

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -862,7 +862,7 @@ namespace ts {
             if (name.kind === SyntaxKind.Identifier) {
                 let message = meaning === SymbolFlags.Namespace ? Diagnostics.Cannot_find_namespace_0 : Diagnostics.Cannot_find_name_0;
 
-                symbol = resolveName(name, (<Identifier>name).text, meaning, message, <Identifier>name);
+                symbol = resolveName(name, (<Identifier>name).text, meaning, !ignoreErrors ? message : undefined, <Identifier>name);
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: i would just have htis as ```ignoreErrors ? undefined : message```

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -3308,6 +3306,15 @@ namespace ts {
             return getSignaturesOfObjectOrUnionType(getApparentType(type), kind);
         }
 
+        function typeHasConstructSignatures(type: Type): boolean {
+            let apparentType = getApparentType(type);
+            if (apparentType.flags & (TypeFlags.ObjectType | TypeFlags.Union)) {
+                let resolved = resolveObjectOrUnionTypeMembers(<ObjectType>type);
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: calling this explicitly seems odd.  Isn't there a way to more simply ask for the resolved type?

File: src/compiler/types.ts [hunk 3]
```diff
@@ -1525,9 +1539,7 @@ namespace ts {
         getConstantValue(node: EnumMember | PropertyAccessExpression | ElementAccessExpression): number;
         getBlockScopedVariableId(node: Identifier): number;
         getReferencedValueDeclaration(reference: Identifier): Declaration;
-        serializeTypeOfNode(node: Node): string | string[];
-        serializeParameterTypesOfNode(node: Node): (string | string[])[];
-        serializeReturnTypeOfNode(node: Node): string | string[];
+        isTypeWithValue(node: TypeReferenceNode): TypeWithValueResolutionResult;
```
Comment (CyrusNajmabadi) [inline src/compiler/types.ts]: 'is' function returning a non-boolean result.  Smells weird.  this seems more like getTypeWithValueKind(...), or something like that.

File: src/compiler/types.ts [hunk 4]
```diff
@@ -1501,6 +1501,20 @@ namespace ts {
     export interface SymbolAccessiblityResult extends SymbolVisibilityResult {
         errorModuleName?: string // If the symbol is not visible from module, module's name
     }
+    
+    /* @internal */
+    export enum TypeWithValueResolutionResult {
+        Unknown,
+        ConstructorTypeWithValue,
```
Comment (CyrusNajmabadi) [inline src/compiler/types.ts]: can you give examples.  esp for this one.

**Comment by SnareChops** [comment]
I would say that this is very high priority and believe that it should go into 1.5. This is a critical bug that breaks any application using the reflection metadata api (yes experimental) and ES6 style exports. For example:

```typescript
export default class FriendsService {
  names: Array<string>;
  constructor() {
    this.names = ["Alice", "Aarav", "Martín", "Shannon", "Ariana", "Kai"];
  }
}
```

```typescript
import FriendsService from './FriendsService';
@Component({
  ...
  appInjector: [FriendsService]
})
class DisplayComponent {
  myName: string;
  names: Array<string>;
  constructor(friendsService: FriendsService) {
    this.myName = 'Alice';
    this.names = friendsService.names;
  }
}
```

This example is taken directly from [angular](https://angular.io/docs/js/latest/guide/displaying-data.html) but I moved the two classes into separate files and used the ES6 `import` and `export` statements.

In this case I would simply try to remove the type annotation on the constructor parameter, but that is required by Angular.... So either way it's broken. Therefore I make the argument that this is a critical bug and should be included in 1.5

**Comment by mhegazy** [comment]
@SnareChops sorry for the delay, it has been a relatively tricky change to get it. we will get you a build to unblock you asap. as for the release vehicle, i do not think 1.5 is an option. but we are working on getting a nightly release in place, so should have an "official" release with the fix soon.

File: src/compiler/checker.ts (hunk 2 above)
Comment (rbuckton) [inline src/compiler/checker.ts]: This is what the existing `typeHasCallOrCosntructSignatures` does below, this just checks for construct signatures only.

**Pull request merged**